# Python basics

Install Python on your own computer.

Run jupyter notebook. Familiarize yourself with the main elements of the GUI.

## Python as a pocket calculator

In [1]:
2 + 3

5

In [2]:
2*3

6

In [3]:
2/3  # real (not integer) division

0.6666666666666666

In [4]:
2**3

8

Python’s math module, in the standard library, can help you work on math-related problems in code. It contains many useful functions, such as remainder() and factorial(). It also includes the Python square root function, sqrt().

In [7]:
import math
math.sqrt(2)

1.4142135623730951

In [8]:
math.pi  # this is a permanent (predefined) variable

3.141592653589793

In [9]:
math.sin(math.pi/2)

1.0

To obtain help on a certain function, you can use

In [12]:
?math.sin

In [15]:
 math.sin?

## Operator precedence

From Python documentation on [operator precedence](http://docs.python.org/reference/expressions.html)

1. Parentheses (grouping)
2. Exponentiation $**$
3. Multiplication, division, remainder $*, /, \%$
4. Addition, subtraction

Note that expressions are always evaluated from left to right!

Examples: Try to guess the results first without using Python:

In [17]:
1 + 2 * 3

7

In [18]:
4 / 2 * 2

4.0

In [19]:
2 * 2 ** 3

16

In [25]:
(2 ** 2) ** 3

64

In [22]:
2 ** (1 + 2) / 3

2.6666666666666665

Now the other way around - type the following mathematical expressions in Python:

$$\frac{1}{2*3}$$

$$\left(2^{2}\right)^{3}$$

$${2^{2}}^{2}$$

## Representation of numbers in Python

Python includes three numeric types to represent numbers: integers, float, and complex number.

### Int

In Python, integers are zero, positive or negative whole numbers without a fractional part and having unlimited precision, e.g. 0, 100, -10. The followings are valid integer literals in Python.

In [27]:
0

0

In [28]:
100

100

In [29]:
-10

-10

In [30]:
1234567890

1234567890

In [32]:
y=5000000000000000000000000000000000000000000000000000000
y

5000000000000000000000000000000000000000000000000000000

Integers can be binary, octal, and hexadecimal values.

In [33]:
0b11011000 # binary

216

In [34]:
0o12 # octal

10

In [35]:
0x12 # hexadecimal

18

All integers literals or variables are objects of the int class. Use the type() method to get the class name, as shown below.

In [36]:
type(100)

int

In [37]:
x=1234567890
type(x)

int

In [38]:
type(y)

int

Leading zeros in non-zero integers are not allowed e.g. 000123 is invalid number, 0000 is 0.

In [39]:
x=01234567890

SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers (3636360402.py, line 1)

Python does not allow comma as number delimiter. Use underscore _ as a delimiter instead.

In [40]:
x=1_234_567_890
x

1234567890

Note that integers must be without a fractional part (decimal point). It it includes a fractional then it becomes a float.

In [41]:
x=5
type(x)

int

In [42]:
x=5.0
type(x)

float

#### Binary

A Number having 0b with eight digits in the combination of 0 and 1 represent the binary numbers in Python. For example, 0b11011000 is a binary number equivalent to integer 216.

In [43]:
x=0b11011000
x

216

In [44]:
x=0b_1101_1000
x

216

In [45]:
type(x)

int

#### Octal

A number having 0o or 0O as prefix represents an octal number. For example, 0O12 is equivalent to integer 10.

In [46]:
x=0o12
x

10

In [47]:
type(x)

int

### Hexadecimal

A number with 0x or 0X as prefix represents hexadecimal number. For example, 0x12 is equivalent to integer 18.

In [48]:
x=0x12
x

18

In [49]:
type(x)

int

### Float

In Python,  floating point numbers (float) are positive and negative real numbers with a fractional part denoted by the decimal symbol . or the scientific notation E or e, e.g. 1234.56, 3.142, -1.55, 0.23.

In [51]:
f=1.2
f

1.2

In [52]:
type(f)

float

Floats can be separated by the underscore _, e.g. 123_42.222_013 is a valid float.

In [53]:
f=123_42.222_013
f

12342.222013

The maximum size depends on your system. The float beyond its maximum size referred as "inf", "Inf", "INFINITY", or "infinity". Float 2e400 will be considered as infinity for most systems.

In [54]:
f=2e400
f

inf

Scientific notation is used as a short representation to express floats having many digits. For example: 345.56789 is represented as 3.4556789e2 or 3.4556789E2

In [55]:
f=1e3
f

1000.0

In [56]:
f=1e5
f

100000.0

In [57]:
f=3.4556789e2
f

345.56789

Use the float() function to convert string, int to float.

In [58]:
float('5.5')

5.5

In [59]:
float('5')

5.0

In [60]:
float('     -5')

-5.0

In [61]:
float('1e3')

1000.0

In [62]:
float('-Infinity')

-inf

In [63]:
float('inf')

inf

### Complex Numbers

A complex number is a number with real and imaginary components. For example, 5 + 6j is a complex number where 5 is the real component and 6 multiplied by j is an imaginary component

In [64]:
a=5+2j
a

(5+2j)

In [65]:
type(a)

complex

You must use j or J as imaginary component. Using other character will throw syntax error.

In [66]:
a=5+2k

SyntaxError: invalid syntax (336981122.py, line 1)